zebra-chain/src/orchard/tree.rs

```diff
@@ -228,6 +261,49 @@ impl NoteCommitmentTree {
     pub fn uncommitted() -> pallas::Base {
         pallas::Base::one().double()
     }
+
+    /// Count of note commitments added to the tree.
+    ///
+    /// For Orchard, the tree is capped at 2^32.
+    pub fn count(&self) -> u32 {
+        self.parents.iter().enumerate().fold(
+            match (self.left, self.right) {
+                (None, None) => 0,
+                (Some(_), None) => 1,
+                (Some(_), Some(_)) => 2,
+                (None, Some(_)) => unreachable!(),
+            },
+            |acc, (i, p)| {
+                // Treat occupation of parents array as a binary number
+                // (right-shifted by 1)
+                acc + if p.is_some() { 1 << (i + 1) } else { 0 }
+            },
+        )
+    }
+
+    fn is_complete(&self) -> bool {
+        self.left.is_some()
+            && self.right.is_some()
+            && self.parents.len() == MERKLE_DEPTH - 1
+            && self.parents.iter().all(|p| p.is_some())
+    }
+}
+
+impl From<Vec<pallas::Base>> for NoteCommitmentTree {
+    /// Compute the tree from a whole bunch of note commitments at once.
+    fn from(values: Vec<pallas::Base>) -> Self {
+        let mut tree = Self::default();
+
+        if values.is_empty() {
+            return tree;
+        }
+
+        for cm_x in values {
+            let _ = tree.append(cm_x);
+        }
+
+        tree
+    }
 }
 
 // TODO: check empty roots, incremental roots, as part of https://github.com/ZcashFoundation/zebra/issues/1287
```

Comment:
Test coverage: We seem to be doing a lot more tests for Sapling than Orchard.

Can we do similar tests for Orchard? Or is that another ticket?

Reply:
We can use test vectors from zcashd (`src/gtest/test_merkletree.cpp` / `src/test/data/merkle_roots_orchard.h`) which are from https://github.com/zcash-hackworks/zcash-test-vectors/pull/39/files#diff-53efbb4ad196b84742420f65d1ef919a9dc4385bb0f2cb2b8903031a003886a3